api/resmap/reswrangler.go

```diff
@@ -26,6 +26,10 @@ type resWrangler struct {
 	// specify in kustomizations to be maintained and
 	// available as an option for final YAML rendering.
 	rList []*resource.Resource
+
+	// Map of used ids in the rList, to avoid recalculating
+	// resource ids and to get quick duplicate checks in Append
+	ids map[resid.ResId]bool
```

Comment:
I can see the appeal of adding an extra map to make the lookup faster, but I'm concerned that this implementation is too brittle. How can we ensure that this is always up to date?

One example I've found is that this PR does not update `ids` in `resWrangler.ApplyFilter`: https://github.com/kubernetes-sigs/kustomize/blob/e8b521e3ab8c5184ee7f895fff156642b0a63db6/api/resmap/reswrangler.go#L762. Here, `rList` is modified, but `ids` is not. 

It would improve our confidence in this change if you can edit the tests in `reswrangler_test.go` in such a way that makes it obvious that the `ids` here are properly being updated. This would be required for every public function of `resWrangler`.

Reply:
Thanks for your feedback. I agree, I also have another concern: are IDs allowed to change by mutating the resources in a resmap (without notifying resmap)?

I was thinking thinking of a couple of possible approaches:
* extract core "ordered map" functions to a new generic small API, and use this in reswrangler
* probably simpler: replace the rList slice with rList struct{ id, element }. It would reduce the CurId calls but not improve lookup complexity

Reply:
If you want to define an "ordered map" package (your first option), I think that would be helpful too - and I can think of a few other places that it would be useful - but defining the ordered map API should probably be its own PR (and then another PR to use in reswrangler).

Since that option is heavier weight and more work on your side, I'd be fine with starting with your second option in this PR if you'd like.

Reply:
FYI, I took another look at the source code, and this approach does not work, as each resource might be modified without the resmap knowing about it.

I think the resmap needs to be informed about all mutations to the objects to ensure the CurId is kept up to date. I'll propose some alternatives for this later when I have more time.

Reply:
+1 for using a generic ordered map API.